File: WorkManagerJava/app/src/main/java/com/websarva/wings/android/workmanagerjava/ThreeStatesWorker.java
package com.websarva.wings.android.workmanagerjava;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

public class ThreeStatesWorker extends Worker {
	public ThreeStatesWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
		super(context, workerParams);
	}

	@NonNull
	@Override
	public Result doWork() {
		Result returnVal = Result.success();
		long failureCount = Math.round(Math.random() * 20);
		long retryCount = Math.round(Math.random() * 20);
		Log.i("ThreeStatesWorker", "失敗と判定する数値: " + failureCount);
		Log.i("ThreeStatesWorker", "リトライと判定する数値: " + retryCount);
		for(int i = 1; i <= 10; i++) {
			Log.i("ThreeStatesWorker", "現在のカウント: " + i);
			if(i == failureCount) {
				returnVal = Result.failure();
				break;
			}
			else if(i == retryCount) {
				returnVal = Result.retry();
				break;
			}
			else {
				try {
					Thread.sleep(1000);
				} catch (InterruptedException e) {
					returnVal = Result.failure();
				}
			}
		}
		return returnVal;
	}
}

File: WorkManagerKotlin/app/src/main/java/com/websarva/wings/android/workmanagerkotlin/ThreeStatesWorker.kt
package com.websarva.wings.android.workmanagerkotlin

import android.content.Context
import android.util.Log
import androidx.work.Worker
import androidx.work.WorkerParameters

class ThreeStatesWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {
	override fun doWork(): Result {
		var returnVal = Result.success()
		val failureCount = (1..20).random()
		val retryCount = (1..20).random()
		Log.i("ThreeStatesWorker", "失敗と判定する数値: ${failureCount}")
		Log.i("ThreeStatesWorker", "リトライと判定する数値: ${retryCount}")
		for(i in 1..25) {
			Log.i("ThreeStatesWorker", "現在のカウント: ${i}")
			if(i == failureCount) {
				returnVal = Result.failure()
				break
			}
			else if(i == retryCount) {
				returnVal = Result.retry()
				break
			}
			else {
			Thread.sleep(1000)
			}
		}
		return returnVal
	}
}
